tests/integration_tests/data_steward/cdr_cleaner/cleaning_rules/cleaning_tests_base.py

```diff
@@ -0,0 +1,229 @@
+            cls.client = bq.get_client(cls.project_id)
+
+            # the base class will try to determine this from the list of
+            # tables_and_counts if the list is empty
+            if not cls.fq_table_names:
+                for table_info in cls.tables_and_counts:
+                    cls.fq_table_names.append(
+                        table_info.get('fq_table_name', ''))
+
+            table_info = {
+                # list of fully qualified table names to create.
+                # will be created with project defined schemas
+                'fq_table_names': cls.fq_table_names,
+                'project_id': cls.project_id,
+                'exists_ok': True,
+            }
+            bq.create_tables(**table_info)
+
+        @classmethod
+        def tearDownClass(cls):
+            """
+            Remove the test dataset table(s).
+            """
+            for table in cls.fq_table_names + cls.fq_sandbox_table_names:
+                cls.client.delete_table(table)
+
+        def drop_rows(self, fq_table_name):
+            """
+            Helper function to drop table rows and assert the drop finished.
+
+            Raises an assertion error if the table records are not dropped.
+            Drops all rows where the {table}_id field > -1.  So all loaded test
+            data should ensure {table}_id fields are set to positive numbers.
+
+            :param fq_table_name: a fully qualified table name to drop all
+                records for
+            """
+            table_name = ''
+            for info in self.tables_and_counts:
+                if info.get('fq_table_name', '') == fq_table_name or \
+                    info.get('fq_sandbox_table_name', '') == fq_table_name:
+                    table_name = info.get('name')
+
+            query = f"delete from {fq_table_name} where {table_name}_id > -1"
+
+            response = bq.query(query, self.project_id)
+
+            # start the job and wait for it to complete
+            self.assertIsNotNone(response.result())
+            self.assertIsNone(response.exception())
+
+        def tearDown(self):
+            """
+            Ensure the data is dropped from the table(s).
+            """
+            for table in self.fq_table_names + self.fq_sandbox_table_names:
+                self.drop_rows(table)
+
+        def setUp(self):
+            """
+            Add data to the tables for the rule to run on.
+            """
+            # load the data from the sql strings and validate it loaded
+            for query in self.sql_load_statements:
+                response = bq.query(query, self.project_id)
+                self.assertIsNotNone(response.result())
+                self.assertIsNone(response.exception())
+
+    class DropRowsTestBase(CleaningRulesTestBase):
+        """
+        Class that can be extended and used to test cleaning rules that drop row data.
+        """
+
+        def test_execution_of_get_query_specs(self):
```

Comment:
This feels like it's forcing a single test case for the test class, which might be overly restrictive. I might just make this whole thing a helper method which would be invoked by a more specifically name subclass test, e.g. rename this to `run_test_execution_of_get_query_specs({... expectations ... })` and move the test output expectations into the parameter rather than on the base class definition

Reply:
No.  You're not thinking about this from an OOP perspective.  I can rename the method to `test_execution_default`, but I still think this should be a test because it is a **default** method.  Meaning, if you don't declare any tests, this will execute for a properly set up extending class.  If you don't like what this method is doing, you can override this method in your extending class with a new definition (even if it's just a `pass` body.  But at that point, I probably wouldn't extend this class).  If you like what this method is doing and want to add additional tests, you can define the additional tests in your extending class.